python-frontend/src/main/java/org/sonar/python/tree/CallExpressionImpl.java

```diff
@@ -187,24 +189,46 @@ public List<Expression> typeDependencies() {
 
   @Override
   public PythonType typeV2() {
-    if (callee().typeV2() instanceof ClassType classType) {
-      return new ObjectType(classType);
+    TypeSource typeSource = computeTypeSource();
+    PythonType pythonType = returnTypeOfCall(callee().typeV2());
+    return pythonType != PythonType.UNKNOWN ? new ObjectType(pythonType, typeSource) : PythonType.UNKNOWN;
+  }
+
+  static PythonType returnTypeOfCall(PythonType calleeType) {
+    if (calleeType instanceof ClassType classType) {
+      return classType;
     }
-    if (callee().typeV2() instanceof FunctionType functionType) {
-      return functionType.returnType();
+    if (calleeType instanceof FunctionType functionType) {
+      PythonType returnType = functionType.returnType();
+      if (returnType.equals(PythonType.UNKNOWN)) {
+        return PythonType.UNKNOWN;
+      }
+      return returnType;
     }
-    if (callee().typeV2() instanceof UnionType unionType) {
-      PythonType result = null;
+    if (calleeType instanceof UnionType unionType) {
+      Set<PythonType> types = new HashSet<>();
       for (PythonType candidate : unionType.candidates()) {
         if (candidate instanceof ClassType classType) {
```

Comment:
why can't you instead this conditional checking for instanceof call:
```java
for (var candidate : unionType.candidates()) {
  var candidateReturnType = returnTypeOfCall(candidate)
  ...
}
```

Reply:
Yes indeed, thanks! I now realize I missed your initial point about recursion, sorry about that.